Optim foreach cleanup for NAdam

mikaylagawarecki opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70953
* #70484
* #70483
* #70482
* #70481
* #70295
* #70231
* #70230
* __->__ #70229
* #69982
* #69981
* #69980
* #71335
* #71334
* #71333

Add foreach flag to NAdam optimizer + cleanup

Differential Revision: [D33767873](https://our.internmc.facebook.com/intern/diff/D33767873)

mikaylagawarecki committed:
Optim foreach cleanup for NAdam

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for NAdam"




[ghstack-poisoned]

mikaylagawarecki commented on torch/optim/nadam.py:
```diff
@@ -116,23 +122,145 @@ def step(self, closure=None):
+    mus = [beta1 * (1. - 0.5 * (0.96 ** (step * momentum_decay))) for step in state_steps]
+    mu_nexts = [beta1 * (1. - 0.5 * (0.96 ** ((step + 1) * momentum_decay)))
+                for step in state_steps]
+    mu_products = [mu * mu_product for mu, mu_product in zip(mus, mu_products)]
```
The multitensor Nadam class does [an update on mu_product](https://github.com/pytorch/pytorch/blob/master/torch/optim/_multi_tensor/nadam.py#L90-L92) before the call to nadam and then does computations on `mu_products` directly whereas the singletensor class does [an update on mu_product](https://github.com/pytorch/pytorch/blob/master/torch/optim/nadam.py#L132-L136) after the call to nadam and computes the new `mu_products` within the nadam function. In this combined class I preserved the singletensor behavior so I added line 245 here to ensure the same behavior

albanD replied:
This could be done with a foreach op no?
Also the existing code is modifying mu_products inplace, don't we want to preserve that/

albanD replied:
This could be done with a foreach op no?
Also the existing code is modifying mu_products inplace, don't we want to preserve that?

mikaylagawarecki replied:
Hm I didn't do this in place because each `mu_product` is a float rather than a tensor so even if it's done in-place i don't think it updates the underlying `state['mu_product']`, but there is an update step within the NAdam.step function on line 146 which updates `state['mu_product']`, which is preserved from the single tensor version, am I thinking about this correctly?

albanD replied:
Aren't you computing the same thing twice here then?
I am a bit confused now where this value is updated for each case.

In general, I think we want to stay as close to the original code as possible. Even if we have to fold some of the state update in the functional function.

mikaylagawarecki replied:
Yes I agree that the same thing is being computed twice, but that was what was being done initially in torch/optim/nadam.py. The single tensor version does the update of mu_product [after calling F.nadam](https://github.com/pytorch/pytorch/blob/master/torch/optim/nadam.py#L132-L136) (and this same computation is done in the function on lines 224 and 226) whereas the multitensor one does it [before calling F.nadam](https://github.com/pytorch/pytorch/blob/master/torch/optim/_multi_tensor/nadam.py#L90-L92) and so the logic isn't repeated in the function! So I think we need to change the code of either one of the functional forms (for example if I preserve the initial multitensor version and get rid of this change to _multi_tensor_nadam I think I would have to remove line 226 of _single_tensor_nadam (`mu_product = mu_product * mu`), does that make sense?

albanD replied:
The two options I see are:
- We want to keep the exact same behavior as before and so we fold the second update for the single Tensor impl into the functional. And fold the multitensor computation into its functional version.
- We want to remove this duplicate code and so modify one of the two to match the other. In such case, I think we should keep the version that does not do the computation twice as it is better! Also the functional version should perform the full step, so I would argue it is a bug that part of it is done in the optimizer outside of the functional call.

mikaylagawarecki replied:
hm which option do you think I should go ahead with? It seems that ASGD and SGD have the same bug where part of the state is updated outside the functional form in single tensor ASGD and SGD as well. I think they did this (for ASGD and NAdam) because arguments were passed as floats rather than singleton tensors so they couldn't update state within the function.

mikaylagawarecki committed:
Update on "Optim foreach cleanup for NAdam"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for NAdam"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for NAdam"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for NAdam"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for NAdam"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for NAdam"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for NAdam"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for NAdam"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for NAdam"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Optim foreach cleanup for NAdam"




[ghstack-poisoned]

mikaylagawarecki commented on torch/optim/nadam.py:
```diff
@@ -116,23 +126,155 @@ def step(self, closure=None):
+        mu = beta1 * (1. - 0.5 * (0.96 ** (step * momentum_decay)))
+        mu_next = beta1 * (1. - 0.5 * (0.96 ** ((step + 1) * momentum_decay)))
+        mu_product = mu_product * mu
+        mu_product_next = mu_product * mu * mu_next
```
I think this step is actually a bug - `mu_product_next` should just be `mu_product * mu_next`, so the additional step actually did not have to be added on line 259 for parity, should I file an issue for this?

albanD replied:
Yes please file an issue! Thanks for looking into the computation in details.

mikaylagawarecki committed:
Update on "Optim foreach cleanup for NAdam"




[ghstack-poisoned]

albanD reviewed (approved):
Much cleaner!

mikaylagawarecki committed:
Update on "Optim foreach cleanup for NAdam"




[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33767873).

mikaylagawarecki committed:
Update on "Optim foreach cleanup for NAdam"


Differential Revision: [D33767873](https://our.internmc.facebook.com/intern/diff/D33767873)

[ghstack-poisoned]

mikaylagawarecki commented:
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33767873).
